libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -0,0 +1,550 @@
+
+    // for each visibility graph
+    const AP_OAVisGraph* visgraph_ptrs[] = {&_polyfence_visgraph, &_destination_visgraph};
+    for (uint8_t v=0; v<ARRAY_SIZE(visgraph_ptrs); v++) {
+
+        // skip if empty
+        const AP_OAVisGraph &curr_visgraph = *visgraph_ptrs[v];
+        if (curr_visgraph.num_items() == 0) {
+            continue;
+        }
+
+        // search visibility graph for items visible from current_node
+        for (uint8_t i=0; i<curr_visgraph.num_items(); i++) {
+            const AP_OAVisGraph::VisGraphItem &item = curr_visgraph[i];
+            // match if current node's id matches either of the id's in the graph (i.e. either end of the vector)
+            if ((curr_node.id == item.id1) || (curr_node.id == item.id2)) {
+                AP_OAVisGraph::OAItemID matching_id = (curr_node.id == item.id1) ? item.id2 : item.id1;
+                // find item's id in node array
+                node_index item_node_idx;
+                if (find_node_from_id(matching_id, item_node_idx)) {
+                    // if current node's distance + distance to item is less than item's current distance, update item's distance
+                    const float dist_to_item_via_current_node = _short_path_data[curr_node_idx].distance_cm + item.distance_cm;
+                    if (dist_to_item_via_current_node < _short_path_data[item_node_idx].distance_cm) {
+                        // update item's distance and set "distance_from_idx" to current node's index
+                        _short_path_data[item_node_idx].distance_cm = dist_to_item_via_current_node;
+                        _short_path_data[item_node_idx].distance_from_idx = curr_node_idx;
+                    }
+                }
+            }
+        }
+    }
+}
+
+// find a node's index into _short_path_data array from it's id (i.e. id type and id number)
+// returns true if successful and node_idx is updated
+bool AP_OADijkstra::find_node_from_id(const AP_OAVisGraph::OAItemID &id, node_index &node_idx) const
+{
+    // source node is always the first node
+    if ((id.id_type == AP_OAVisGraph::OATYPE_SOURCE) && (_short_path_data_numpoints > 0)) {
+        node_idx = 0;
+        return true;
+    }
+
+    // destination is always the 2nd node
+    if ((id.id_type == AP_OAVisGraph::OATYPE_DESTINATION) && (_short_path_data_numpoints > 1)) {
+        node_idx = 1;
+        return true;
+    }
+
+    // must be a fence node which start from 3rd node
+    if ((id.id_type == AP_OAVisGraph::OATYPE_FENCE_POINT) && (_short_path_data_numpoints > id.id_num + 2)) {
+        node_idx = id.id_num + 2;
+        return true;
+    }
+
+    // could not find node
+    return false;
+}
+
+// find index of node with lowest tentative distance (ignore visited nodes)
+// returns true if successful and node_idx argument is updated
+bool AP_OADijkstra::find_closest_node_idx(node_index &node_idx) const
+{
+    uint16_t lowest_idx = 0;
```

Comment:
You're still mixing types on this though. It's `uint16_t` but you are down casting to `node_index`. These should be consistent if you are going to use the typedefs like this.

Reply:
yes, you're right.  I'll replace the uint16_t with node_index.  txs again